[GFC] Build Layout::ElementBox tree from RenderGrid.

yulunwu2 opened the pull request:
#### c1c037979677818a2d80a6386b4a2d381b8de45e
<pre>
[GFC] Build Layout::ElementBox tree from RenderGrid.
<a href="https://bugs.webkit.org/show_bug.cgi?id=299053">https://bugs.webkit.org/show_bug.cgi?id=299053</a>
&lt;<a href="https://rdar.apple.com/160810280">rdar://160810280</a>&gt;

Reviewed by Sammy Gill.

This PR implements BoxTreeUpdater::buildTreeForGridContent() to convert a
RenderGrid and its children into a tree of Layout::ElementBox and populate
LayoutIntegrationGridLayout&apos;s m_gridBox.

This PR also adds a utility function to LayoutIntegrationGridLayout to support

WTF_ALWAYS_LOG(&lt;LayoutIntegration::GridLayout object&gt;);

This is a QOL function to streamline development and debugging.

Example debug output:

gridLayoutGridLayout@0x16cf378f8 gridBox=0x135310270
  [0] box=0x135310340 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc230
  [1] box=0x135310410 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc510
  [2] box=0x1353104e0 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc680
  [3] box=0x1353105b0 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc7f0
  [4] box=0x135310680 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc960
  [5] box=0x135310750 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fcad0

* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
(WebCore::LayoutIntegration::BoxTreeUpdater::buildTreeForGridContent):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
(WebCore::LayoutIntegration::GridLayout::layout):
(WebCore::LayoutIntegration::operator&lt;&lt;):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h:

Canonical link: <a href="https://commits.webkit.org/300208@main">https://commits.webkit.org/300208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4be3a6e1f8c1eb6138e77fc419d88e9532c6974

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128159 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73732 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/61e5aef9-5088-4af7-89dc-82f6a9d187f5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42055 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92436 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a7029680-c4dc-413d-bf67-1fd25976133f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33557 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108958 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73097 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/22a32073-3458-428b-b167-bf1efedd0904) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71676 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103054 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130957 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48576 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36955 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101009 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105170 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100898 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25593 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46278 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24383 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45277 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48435 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54162 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47904 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51253 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49587 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d69c80f1bb2071a4507e30ac2164c48ca4e47620. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50896

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/97dd92392eab8955f63d13258a20b6f40ae5abab. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50896

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c9680e514457697c6ea5280bac7ff1bf263124b4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50896

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -41,7 +45,25 @@ GridLayout::GridLayout(RenderGrid& renderGrid)
+        stream << "\n  [" << index++ << "] box=" << static_cast<const void*>(box.ptr());
+        stream << " anonymous=" << (box->isAnonymous() ? "yes" : "no");
+        stream << " establishesContext=" << (box->establishesFormattingContext() ? "yes" : "no");
+        stream << " display=" << static_cast<unsigned>(box->style().display());
```
Seems like we have a text stream operator for `DisplayType` in `RenderStyleConstants.h` so could this just be 
```
 stream << " display=" << box->style().display();
```

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -41,7 +45,25 @@ GridLayout::GridLayout(RenderGrid& renderGrid)
+    stream << " gridBox=" << &layout.gridBox();
+    size_t index = 0;
+    for (CheckedRef<const Layout::Box> box : Layout::formattingContextBoxes(layout.gridBox())) {
+        stream << "\n  [" << index++ << "] box=" << static_cast<const void*>(box.ptr());
```
What's the purpose of the static cast? Shouldn't this already log the address?

yulunwu2 replied:
Done, thanks for the review!

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -26,8 +26,12 @@
 #include "LayoutIntegrationBoxTreeUpdater.h"
 #include "RenderGrid.h"
+#include <wtf/CheckedPtr.h>
+#include <wtf/CheckedRef.h>
```
You don't need this one since it's already included in `CheckedPtr.h`

yulunwu2 replied:
I think (?) its preferable to not rely on chained includes to have all the deps needed.

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h:
```diff
@@ -44,10 +48,10 @@ class GridLayout {
     void layout();
 
-private:
     const Layout::ElementBox& gridBox() const { return *m_gridBox; }
     Layout::ElementBox& gridBox() { return *m_gridBox; }
 
+private:
```
Done, thanks for the review!

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -41,7 +45,25 @@ GridLayout::GridLayout(RenderGrid& renderGrid)
+    for (CheckedRef<const Layout::Box> box : Layout::formattingContextBoxes(layout.gridBox())) {
+        stream << "\n  [" << index++ << "] box=" << static_cast<const void*>(box.ptr());
+        stream << " anonymous=" << (box->isAnonymous() ? "yes" : "no");
+        stream << " establishesContext=" << (box->establishesFormattingContext() ? "yes" : "no");
```
IIRC grid items should always establish their own formatting context:

https://www.w3.org/TR/css-grid-1/#grid-item-display

yulunwu2 committed:
[GFC] Build Layout::ElementBox tree from RenderGrid.
https://bugs.webkit.org/show_bug.cgi?id=299053
<rdar://160810280>

Reviewed by Sammy Gill.

This PR implements BoxTreeUpdater::buildTreeForGridContent() to convert a
RenderGrid and its children into a tree of Layout::ElementBox and populate
LayoutIntegrationGridLayout's m_gridBox.

This PR also adds a utility function to LayoutIntegrationGridLayout to support

WTF_ALWAYS_LOG(<LayoutIntegration::GridLayout object>);

This is a QOL function to streamline development and debugging.

Example debug output:

gridLayoutGridLayout@0x16cf378f8 gridBox=0x135310270
  [0] box=0x135310340 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc230
  [1] box=0x135310410 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc510
  [2] box=0x1353104e0 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc680
  [3] box=0x1353105b0 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc7f0
  [4] box=0x135310680 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fc960
  [5] box=0x135310750 anonymous=no establishesContext=yes display=1 renderer=RenderBlock@0x1352fcad0

* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
(WebCore::LayoutIntegration::BoxTreeUpdater::buildTreeForGridContent):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
(WebCore::LayoutIntegration::GridLayout::layout):
(WebCore::LayoutIntegration::operator<<):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h:

Canonical link: https://commits.webkit.org/300208@main
